drivers/counter/ds3231.c

```diff
@@ -0,0 +1,1060 @@
+		return rc;
+	}
+
+	*cp = (struct rtc_ds3231_alarm){ 0 };
+	cp->flags = decode_alarm(rbp, (id == 0), &cp->time);
+	cp->handler = data->alarm_handler[id];
+	cp->user_data = data->alarm_user_data[id];
+
+	return 0;
+}
+
+static int ds3231_counter_cancel_alarm(struct device *dev,
+				       u8_t id)
+{
+	struct ds3231_data *data = dev->driver_data;
+	const struct ds3231_config *cfg = dev->config->config_info;
+
+	if (id >= cfg->generic.channels) {
+		return -EINVAL;
+	}
+
+	data->alarm_handler[id] = NULL;
+	data->alarm_user_data[id] = NULL;
+
+	return sc_ctrl(dev, 0, RTC_DS3231_ALARM1 << id);
+}
+
+
+static int ds3231_set_alarm(struct device *dev,
+			    u8_t id,
+			    const struct rtc_ds3231_alarm *cp)
+{
+	struct ds3231_data *data = dev->driver_data;
+	const struct ds3231_config *cfg = dev->config->config_info;
+	u8_t addr;
+	u8_t len;
+
+	if (id == 0) {
+		addr = offsetof(struct register_map, alarm1);
+		len = sizeof(data->registers.alarm1);
+	} else if (id < cfg->generic.channels) {
+		addr = offsetof(struct register_map, alarm2);
+		len = sizeof(data->registers.alarm2);
+	} else {
+		return -EINVAL;
+	}
+
+	u8_t buf[5] = { addr };
+	int rc = encode_alarm(buf + 1, (id == 0), cp->time, cp->flags);
+
+	if (rc < 0) {
+		return rc;
+	}
+
+	/* @todo resolve race condition: a previously stored alarm may
+	 * trigger between clear of AxF and the write of the new alarm
+	 * control.
+	 */
+	rc = rsc_stat(dev, 0U, (RTC_DS3231_ALARM1 << id));
+	if (rc >= 0) {
+		rc = i2c_write(data->i2c, buf, len + 1, cfg->addr);
+	}
+	if ((rc >= 0)
+	    && (cp->handler != NULL)) {
+		rc = sc_ctrl(dev, RTC_DS3231_ALARM1 << id, 0);
+	}
+	if (rc >= 0) {
+		memmove(&data->registers.sec + addr, buf + 1, len);
+		data->alarm_handler[id] = cp->handler;
+		data->alarm_user_data[id] = cp->user_data;
+		data->alarm_flags[id] = cp->flags;
+		validate_isw_monitoring(dev);
+	}
+
+	return rc;
+}
+
+static int ds3231_check_alarms(struct device *dev)
+{
+	struct ds3231_data *data = dev->driver_data;
+	const struct register_map *rp = &data->registers;
+	u8_t mask = (RTC_DS3231_ALARM1 | RTC_DS3231_ALARM2);
+
+	/* Fetch and clear only the alarm flags that are not
+	 * interrupt-enabled.
+	 */
+	return mask & rsc_stat(dev, 0U, (rp->ctrl & mask) ^ mask);
+}
+
+static int check_handled_alarms(struct device *dev)
+{
+	struct ds3231_data *data = dev->driver_data;
+	const struct register_map *rp = &data->registers;
+	u8_t mask = (RTC_DS3231_ALARM1 | RTC_DS3231_ALARM2);
+
+	/* Fetch and clear only the alarm flags that are
+	 * interrupt-enabled.
+	 */
+	return mask & rsc_stat(dev, 0U, rp->ctrl & mask);
+}
+
+static void counter_alarm_forwarder(struct device *dev,
+				    u8_t id,
+				    u32_t syncclock,
+				    void *ud)
+{
+	struct ds3231_data *data = dev->driver_data;
+	counter_alarm_callback_t handler = data->counter_handler[id];
+	u32_t ticks = data->counter_ticks[id];
+
+	data->counter_handler[id] = NULL;
+	data->counter_ticks[id] = 0;
+	handler(dev, id, ticks, ud);
```

Comment:
is it possible that alarm is canceled and callback is still triggered? in that case `handler` may be null,

Reply:
checked at call site; thanks.